source/1_engine/rex_engine/src/memory/memory_tracking.cpp

```diff
@@ -17,27 +23,79 @@ namespace rex
 
   rsl::array<MemoryTag, g_max_allowed_mem_tags>& thread_local_memory_tag_stack()
   {
-    thread_local static rsl::array<MemoryTag, g_max_allowed_mem_tags> stack = {MemoryTag::Global};
+    thread_local static rsl::array<MemoryTag, g_max_allowed_mem_tags> stack = { MemoryTag::Global };
     return stack;
   }
 
+  class AllocationCallStack
+  {
+  public:
+    AllocationCallStack(CallStack callstack, card64 size)
+      : m_callstack(callstack)
+      , m_size(size)
+      , m_ref_count(1)
+    {
+    }
+
+    void add_size(card64 size)
+    {
+      m_size += size;
+      ++m_ref_count;
+    }
+    void sub_size(card64 size)
+    {
+      m_size -= size;
+      --m_ref_count;
+    }
+
+    rsl::memory_size size() const
+    {
+      return m_size;
+    }
+
+    card32 ref_count() const
+    {
+      return m_ref_count;
+    }
+
+  private:
+    CallStack m_callstack;
+    rsl::memory_size m_size;
+    card32 m_ref_count;
+  };
+
+  struct AllocationInfo
+  {
+    AllocationCallStack allocation_callstack;
+    DebugVector<CallStack> deleter_callstacks;
+  };
+
+  // stores the headers for all allocations
   auto& allocation_headers()
   {
-    static UntrackedAllocator allocator {};
+    static UntrackedAllocator allocator{};
     static DebugAllocator dbg_alloc(allocator); // NOLINT(misc-const-correctness)
-    static rsl::vector<MemoryHeader*, decltype(dbg_alloc)> alloc_headers(dbg_alloc);
+    static DebugVector<MemoryHeader*> alloc_headers(dbg_alloc);
     return alloc_headers;
   }
 
+  auto& allocation_info_table()
+  {
+    static UntrackedAllocator allocator{};
+    static DebugAllocator dbg_alloc(allocator); // NOLINT(misc-const-correctness)
+    static DebugHashTable<CallStack, AllocationInfo> alloc_info_table(dbg_alloc);
+    return alloc_info_table;
+  }
+
   card32& thread_local_mem_tag_index()
   {
     thread_local static card32 tag = 0;
     return tag;
   }
 
   MemoryTracker::MemoryTracker()
```

Comment:
Can we maybe dump the stats always in a debug build? cuz now you made this entire system but it does not have much usage. If you made it you probably have ideas about what you want it to do could you please also apply those? The only feature I see being used is the "initialize".

Reply:
the memory stats should only be dumped on demand, they shouldn't get dumped every frame or anything like that.
they will be part of an autotest and also for debugging purposes when you want to track memory and where it's going.
another idea is to create a tracking tool which can track memory over a longer time period (days, weeks, ... depending on our storage capabilities)

Reply:
Why do you think they should only be dumped on demand, if I had a difficult run which is hard to reproduce I can at least look at the memory dump that was automatically generated without me having to think about enabling it. Going to mark this as resolved but we should discuss this.